src/System.Memory/src/System/Buffers/ReadOnlyBuffer.cs

```diff
@@ -0,0 +1,377 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Represents a buffer that can read a sequential series of <typeparam name="T">T</typeparam>.
+    /// </summary>
+    public readonly partial struct ReadOnlyBuffer<T>
+    {
+        private const int IndexBitMask = 0x7FFFFFFF;
+
+        private const int MemoryListStartMask = 0;
+        private const int MemoryListEndMask = 0;
+
+        private const int ArrayStartMask = 0;
+        private const int ArrayEndMask = 1 << 31;
+
+        private const int OwnedMemoryStartMask = 1 << 31;
+        private const int OwnedMemoryEndMask = 0;
+
+        private readonly SequencePosition _bufferStart;
+        private readonly SequencePosition _bufferEnd;
+
+
+        /// <summary>
+        /// Returns empty <see cref="ReadOnlyBuffer{T}"/>
+        /// </summary>
+        public static readonly ReadOnlyBuffer<T> Empty = new ReadOnlyBuffer<T>(new T[0]);
+
+        /// <summary>
+        /// Length of the <see cref="ReadOnlyBuffer{T}"/>.
+        /// </summary>
+        public long Length => GetLength(_bufferStart, _bufferEnd);
+
+        /// <summary>
+        /// Determines if the <see cref="ReadOnlyBuffer{T}"/> is empty.
+        /// </summary>
+        public bool IsEmpty => Length == 0;
+
+        /// <summary>
+        /// Determins if the <see cref="ReadOnlyBuffer{T}"/> is a single <see cref="ReadOnlyMemory{T}"/>.
+        /// </summary>
+        public bool IsSingleSegment => _bufferStart.Segment == _bufferEnd.Segment;
+
+        /// <summary>
+        /// Gets <see cref="ReadOnlyMemory{T}"/> from the first segment.
+        /// </summary>
+        public ReadOnlyMemory<T> First
+        {
+            get
+            {
+                TryGetBuffer(_bufferStart, _bufferEnd, out ReadOnlyMemory<T> first, out _);
```

Comment:
Would this call end up doing a struct copy of _bufferStart and _bufferEnd? These are private fields and we are calling an internal method. Why do they need to be passed in at all?

Reply:
Because this method is used not only with this set of private fields.